// @dojo/widgets/combobox/createComboBoxElement.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "./ComboBox"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var ComboBox_1 = require("./ComboBox");
    /**
     * Configures a ComboBox web component
     */
    function createComboBoxElement() {
        return {
            tagName: 'dojo-combo-box',
            widgetConstructor: ComboBox_1.default,
            attributes: [
                {
                    attributeName: 'autoblur',
                    propertyName: 'autoBlur',
                    value: function (value) { return value === 'false' || value === '0' ? false : true; }
                },
                {
                    attributeName: 'clearable',
                    value: function (value) { return value === 'false' || value === '0' ? false : true; }
                },
                {
                    attributeName: 'invalid',
                    value: function (value) { return value === 'false' || value === '0' ? false : true; }
                },
                {
                    attributeName: 'disabled',
                    value: function (value) { return value === 'false' || value === '0' ? false : true; }
                },
                {
                    attributeName: 'openonfocus',
                    propertyName: 'openOnFocus',
                    value: function (value) { return value === 'false' || value === '0' ? false : true; }
                },
                {
                    attributeName: 'readonly',
                    propertyName: 'readOnly',
                    value: function (value) { return value === 'false' || value === '0' ? false : true; }
                },
                {
                    attributeName: 'required',
                    value: function (value) { return value === 'false' || value === '0' ? false : true; }
                },
                {
                    attributeName: 'value'
                }
            ],
            properties: [
                {
                    propertyName: 'results'
                },
                {
                    propertyName: 'inputProperties'
                },
                {
                    propertyName: 'CustomResultItem'
                },
                {
                    propertyName: 'CustomResultMenu'
                },
                {
                    propertyName: 'getResultLabel'
                },
                {
                    propertyName: 'isResultDisabled'
                },
                {
                    propertyName: 'label'
                }
            ],
            events: [
                {
                    propertyName: 'onBlur',
                    eventName: 'blur'
                },
                {
                    propertyName: 'onFocus',
                    eventName: 'focus'
                },
                {
                    propertyName: 'onChange',
                    eventName: 'change'
                },
                {
                    propertyName: 'onRequestResults',
                    eventName: 'requestresults'
                },
                {
                    propertyName: 'onMenuChange',
                    eventName: 'menuchange'
                }
            ]
        };
    }
    exports.default = createComboBoxElement;
    ;
});
