// @dojo/widgets/combobox/ResultItem.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "@dojo/widget-core/WidgetBase", "@dojo/widget-core/mixins/Themeable", "@dojo/widget-core/d", "./styles/comboBox.m.css"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var WidgetBase_1 = require("@dojo/widget-core/WidgetBase");
    var Themeable_1 = require("@dojo/widget-core/mixins/Themeable");
    var d_1 = require("@dojo/widget-core/d");
    var css = require("./styles/comboBox.m.css");
    ;
    exports.ResultItemBase = Themeable_1.ThemeableMixin(WidgetBase_1.WidgetBase);
    var ResultItem = /** @class */ (function (_super) {
        __extends(ResultItem, _super);
        function ResultItem() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        ResultItem.prototype._onMouseDown = function (event) {
            var _a = this.properties, index = _a.index, onMouseDown = _a.onMouseDown;
            onMouseDown(event, index);
        };
        ResultItem.prototype._onMouseEnter = function (event) {
            var _a = this.properties, index = _a.index, onMouseEnter = _a.onMouseEnter;
            onMouseEnter(event, index);
        };
        ResultItem.prototype._onMouseUp = function (event) {
            var _a = this.properties, onMouseUp = _a.onMouseUp, index = _a.index;
            onMouseUp(event, index);
        };
        ResultItem.prototype.renderResult = function (result) {
            var getResultLabel = this.properties.getResultLabel;
            return d_1.v('div', [getResultLabel(result)]);
        };
        ResultItem.prototype.render = function () {
            var _a = this.properties, isDisabled = _a.isDisabled, result = _a.result, selected = _a.selected;
            return d_1.v('div', {
                'aria-selected': selected ? 'true' : 'false',
                'aria-disabled': isDisabled(result) ? 'true' : 'false',
                classes: this.classes(css.option, selected ? css.selected : null, isDisabled(result) ? css.disabledOption : null),
                'data-selected': selected ? 'true' : 'false',
                onmousedown: this._onMouseDown,
                onmouseenter: this._onMouseEnter,
                onmouseup: this._onMouseUp,
                role: 'option'
            }, [
                this.renderResult(result)
            ]);
        };
        ResultItem = __decorate([
            Themeable_1.theme(css)
        ], ResultItem);
        return ResultItem;
    }(exports.ResultItemBase));
    exports.default = ResultItem;
});
