// @dojo/widgets/combobox/ResultMenu.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "@dojo/widget-core/WidgetBase", "@dojo/widget-core/mixins/Themeable", "@dojo/widget-core/d", "./ResultItem", "./styles/comboBox.m.css"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var WidgetBase_1 = require("@dojo/widget-core/WidgetBase");
    var Themeable_1 = require("@dojo/widget-core/mixins/Themeable");
    var d_1 = require("@dojo/widget-core/d");
    var ResultItem_1 = require("./ResultItem");
    var css = require("./styles/comboBox.m.css");
    ;
    exports.ResultMenuBase = Themeable_1.ThemeableMixin(WidgetBase_1.WidgetBase);
    var ResultMenu = /** @class */ (function (_super) {
        __extends(ResultMenu, _super);
        function ResultMenu() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        ResultMenu.prototype.renderResults = function (results) {
            return results;
        };
        ResultMenu.prototype.render = function () {
            var _a = this.properties, _b = _a.CustomResultItem, CustomResultItem = _b === void 0 ? ResultItem_1.default : _b, getResultLabel = _a.getResultLabel, id = _a.id, _c = _a.isResultDisabled, isResultDisabled = _c === void 0 ? function () { return false; } : _c, onResultMouseDown = _a.onResultMouseDown, onResultMouseEnter = _a.onResultMouseEnter, onResultMouseUp = _a.onResultMouseUp, results = _a.results, selectedIndex = _a.selectedIndex, _d = _a.theme, theme = _d === void 0 ? {} : _d;
            var resultElements = this.renderResults(results.map(function (result, i) { return d_1.w(CustomResultItem, {
                getResultLabel: getResultLabel,
                index: i,
                isDisabled: isResultDisabled,
                key: String(i),
                onMouseDown: onResultMouseDown,
                onMouseEnter: onResultMouseEnter,
                onMouseUp: onResultMouseUp,
                result: result,
                selected: i === selectedIndex,
                theme: theme
            }); }));
            return d_1.v('div', {
                classes: this.classes(css.dropdown),
                id: id,
                role: 'listbox'
            }, resultElements);
        };
        ResultMenu = __decorate([
            Themeable_1.theme(css)
        ], ResultMenu);
        return ResultMenu;
    }(exports.ResultMenuBase));
    exports.default = ResultMenu;
});
