// @dojo/widgets/select/SelectOption.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "@dojo/widget-core/WidgetBase", "@dojo/widget-core/mixins/Themeable", "@dojo/widget-core/d", "./styles/select.m.css"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var WidgetBase_1 = require("@dojo/widget-core/WidgetBase");
    var Themeable_1 = require("@dojo/widget-core/mixins/Themeable");
    var d_1 = require("@dojo/widget-core/d");
    var css = require("./styles/select.m.css");
    exports.SelectOptionBase = Themeable_1.ThemeableMixin(WidgetBase_1.WidgetBase);
    var SelectOption = (function (_super) {
        __extends(SelectOption, _super);
        function SelectOption() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        SelectOption.prototype._onMouseDown = function () {
            var onMouseDown = this.properties.onMouseDown;
            onMouseDown && onMouseDown();
        };
        SelectOption.prototype._onClick = function (event) {
            var _a = this.properties, index = _a.index, onClick = _a.onClick;
            onClick && onClick(event, index);
        };
        SelectOption.prototype.renderLabel = function () {
            return this.properties.optionData.label;
        };
        SelectOption.prototype.render = function () {
            var _a = this.properties, focused = _a.focused, optionData = _a.optionData;
            var optionClasses = [
                css.option,
                focused ? css.focused : null,
                optionData.selected ? css.selected : null,
                optionData.disabled ? css.disabledOption : null
            ];
            return d_1.v('div', {
                role: 'option',
                id: optionData.id,
                classes: this.classes.apply(this, optionClasses),
                'aria-disabled': optionData.disabled ? 'true' : null,
                'aria-selected': optionData.selected + '',
                onclick: this._onClick,
                onmousedown: this._onMouseDown
            }, [this.renderLabel()]);
        };
        return SelectOption;
    }(exports.SelectOptionBase));
    SelectOption = __decorate([
        Themeable_1.theme(css)
    ], SelectOption);
    exports.default = SelectOption;
});
