// @dojo/widgets/textinput/TextInput.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "@dojo/widget-core/WidgetBase", "@dojo/widget-core/mixins/Themeable", "@dojo/widget-core/d", "../label/Label", "./styles/textinput.m.css"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var WidgetBase_1 = require("@dojo/widget-core/WidgetBase");
    var Themeable_1 = require("@dojo/widget-core/mixins/Themeable");
    var d_1 = require("@dojo/widget-core/d");
    var Label_1 = require("../label/Label");
    var css = require("./styles/textinput.m.css");
    exports.TextInputBase = Themeable_1.ThemeableMixin(WidgetBase_1.WidgetBase);
    var TextInput = /** @class */ (function (_super) {
        __extends(TextInput, _super);
        function TextInput() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        TextInput.prototype._onBlur = function (event) { this.properties.onBlur && this.properties.onBlur(event); };
        TextInput.prototype._onChange = function (event) { this.properties.onChange && this.properties.onChange(event); };
        TextInput.prototype._onClick = function (event) { this.properties.onClick && this.properties.onClick(event); };
        TextInput.prototype._onFocus = function (event) { this.properties.onFocus && this.properties.onFocus(event); };
        TextInput.prototype._onInput = function (event) { this.properties.onInput && this.properties.onInput(event); };
        TextInput.prototype._onKeyDown = function (event) { this.properties.onKeyDown && this.properties.onKeyDown(event); };
        TextInput.prototype._onKeyPress = function (event) { this.properties.onKeyPress && this.properties.onKeyPress(event); };
        TextInput.prototype._onKeyUp = function (event) { this.properties.onKeyUp && this.properties.onKeyUp(event); };
        TextInput.prototype._onMouseDown = function (event) { this.properties.onMouseDown && this.properties.onMouseDown(event); };
        TextInput.prototype._onMouseUp = function (event) { this.properties.onMouseUp && this.properties.onMouseUp(event); };
        TextInput.prototype._onTouchStart = function (event) { this.properties.onTouchStart && this.properties.onTouchStart(event); };
        TextInput.prototype._onTouchEnd = function (event) { this.properties.onTouchEnd && this.properties.onTouchEnd(event); };
        TextInput.prototype._onTouchCancel = function (event) { this.properties.onTouchCancel && this.properties.onTouchCancel(event); };
        TextInput.prototype.render = function () {
            var _a = this.properties, controls = _a.controls, describedBy = _a.describedBy, disabled = _a.disabled, invalid = _a.invalid, label = _a.label, maxLength = _a.maxLength, minLength = _a.minLength, name = _a.name, placeholder = _a.placeholder, readOnly = _a.readOnly, required = _a.required, _b = _a.type, type = _b === void 0 ? 'text' : _b, value = _a.value;
            var stateClasses = [
                disabled ? css.disabled : null,
                invalid ? css.invalid : null,
                invalid === false ? css.valid : null,
                readOnly ? css.readonly : null,
                required ? css.required : null
            ];
            var textinput = d_1.v('div', { classes: this.classes(css.inputWrapper) }, [
                d_1.v('input', {
                    classes: this.classes(css.input),
                    'aria-controls': controls,
                    'aria-describedby': describedBy,
                    disabled: disabled,
                    'aria-invalid': invalid ? 'true' : null,
                    maxlength: maxLength ? "" + maxLength : null,
                    minlength: minLength ? "" + minLength : null,
                    name: name,
                    placeholder: placeholder,
                    readOnly: readOnly,
                    'aria-readonly': readOnly ? 'true' : null,
                    required: required,
                    type: type,
                    value: value,
                    onblur: this._onBlur,
                    onchange: this._onChange,
                    onclick: this._onClick,
                    onfocus: this._onFocus,
                    oninput: this._onInput,
                    onkeydown: this._onKeyDown,
                    onkeypress: this._onKeyPress,
                    onkeyup: this._onKeyUp,
                    onmousedown: this._onMouseDown,
                    onmouseup: this._onMouseUp,
                    ontouchstart: this._onTouchStart,
                    ontouchend: this._onTouchEnd,
                    ontouchcancel: this._onTouchCancel
                })
            ]);
            var textinputWidget;
            if (label) {
                textinputWidget = d_1.w(Label_1.default, {
                    extraClasses: { root: Label_1.parseLabelClasses(this.classes.apply(this, [css.root].concat(stateClasses))()) },
                    label: label,
                    theme: this.properties.theme
                }, [textinput]);
            }
            else {
                textinputWidget = d_1.v('div', {
                    classes: this.classes.apply(this, [css.root].concat(stateClasses))
                }, [textinput]);
            }
            return textinputWidget;
        };
        TextInput = __decorate([
            Themeable_1.theme(css)
        ], TextInput);
        return TextInput;
    }(exports.TextInputBase));
    exports.default = TextInput;
});
