diff --git a/examples/editor.js b/examples/editor.js
index e11f96f..4b952b2 100644
--- a/examples/editor.js
+++ b/examples/editor.js
@@ -48,7 +48,7 @@ var __generator = (this && this.__generator) || function (thisArg, body) {
     var project_1 = require("../project");
     var Runner_1 = require("../Runner");
     /* path to the project directory */
-    var PROJECT_DIRECTORY = '../../../projects/';
+    var PROJECT_DIRECTORY = '../projects/';
     /* get references to key DOM nodes */
     var divFile = document.getElementById('div-file');
     var loadProjectButton = document.getElementById('load-project');
