diff --git a/examples/index.js b/examples/index.js
index c272782..9b075a5 100644
--- a/examples/index.js
+++ b/examples/index.js
@@ -1,10 +1,10 @@
 require.config({
     paths: {
-        'vs': '../../../node_modules/monaco-editor/min/vs',
-        '@dojo': '../../../node_modules/@dojo'
+        'vs': '../node_modules/monaco-editor/min/vs',
+        '@dojo': '../node_modules/@dojo'
     },
     packages: [
-        { name: 'src', location: '../../..' }
+        { name: 'src', location: '..' }
     ]
 });
 require(['./editor'], function () { });
