diff --git a/examples/index.js b/examples/index.js
index 02f4aef..a38b2b6 100644
--- a/examples/index.js
+++ b/examples/index.js
@@ -63,7 +63,7 @@ var __generator = (this && this.__generator) || function (thisArg, body) {
     var icons_1 = require("../support/icons");
     var theme_1 = require("../themes/dark/theme");
     /* path to the project directory */
-    var PROJECT_DIRECTORY = '../../../projects/';
+    var PROJECT_DIRECTORY = '../projects/';
     var icons;
     var iconsSourcePath = '../../extensions/vscode-material-icon-theme/out/src/material-icons.json';
     /**