diff --git a/examples/index.js b/examples/index.js
index bc8a605..ab69913 100644
--- a/examples/index.js
+++ b/examples/index.js
@@ -63,9 +63,9 @@ var __generator = (this && this.__generator) || function (thisArg, body) {
     var icons_1 = require("../support/icons");
     var theme_1 = require("../themes/dark/theme");
     /* path to the project directory */
-    var PROJECTS_PATH = '../../../projects/';
-    var EDITOR_THEME = '../../data/editor-dark.json';
-    var iconsSourcePath = '../../extensions/vscode-material-icon-theme/out/src/material-icons.json';
+    var PROJECTS_PATH = '../projects/';
+    var EDITOR_THEME = '../data/editor-dark.json';
+    var iconsSourcePath = '../extensions/vscode-material-icon-theme/out/src/material-icons.json';
     var icons;
     /**
      * An example application widget that incorporates both the Editor and Runner widgets into a simplistic UI