// @dojo/widget-core/Container.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "./WidgetBase", "./d", "./Injector"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var WidgetBase_1 = require("./WidgetBase");
    var d_1 = require("./d");
    var Injector_1 = require("./Injector");
    function Container(component, name, _a) {
        var _b = _a === void 0 ? Injector_1.defaultMappers : _a, _c = _b.getProperties, getProperties = _c === void 0 ? Injector_1.defaultMappers.getProperties : _c, _d = _b.getChildren, getChildren = _d === void 0 ? Injector_1.defaultMappers.getChildren : _d;
        return (function (_super) {
            __extends(class_1, _super);
            function class_1() {
                return _super !== null && _super.apply(this, arguments) || this;
            }
            class_1.prototype.render = function () {
                var _a = this, properties = _a.properties, children = _a.children;
                return d_1.w(name, {
                    bind: this,
                    render: function () { return d_1.w(component, properties, children); },
                    getProperties: getProperties,
                    properties: properties,
                    getChildren: getChildren,
                    children: children
                });
            };
            return class_1;
        }(WidgetBase_1.WidgetBase));
    }
    exports.Container = Container;
    exports.default = Container;
});
