// @dojo/widget-core/d.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "@dojo/core/lang", "@dojo/shim/Symbol", "maquette"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var lang_1 = require("@dojo/core/lang");
    var Symbol_1 = require("@dojo/shim/Symbol");
    var maquette_1 = require("maquette");
    /**
     * The symbol identifier for a WNode type
     */
    exports.WNODE = Symbol_1.default('Identifier for a WNode.');
    /**
     * The symbol identifier for a HNode type
     */
    exports.HNODE = Symbol_1.default('Identifier for a HNode.');
    /**
     * Helper function that returns true if the `DNode` is a `WNode` using the `type` property
     */
    function isWNode(child) {
        return Boolean(child && (typeof child !== 'string') && child.type === exports.WNODE);
    }
    exports.isWNode = isWNode;
    /**
     * Helper function that returns true if the `DNode` is a `HNode` using the `type` property
     */
    function isHNode(child) {
        return Boolean(child && (typeof child !== 'string') && child.type === exports.HNODE);
    }
    exports.isHNode = isHNode;
    function decorate(dNodes, modifier, predicate) {
        var nodes = Array.isArray(dNodes) ? dNodes.slice() : [dNodes];
        while (nodes.length) {
            var node = nodes.pop();
            if (node) {
                if (!predicate || predicate(node)) {
                    modifier(node);
                }
                if ((isWNode(node) || isHNode(node)) && node.children) {
                    nodes = nodes.concat(node.children);
                }
            }
        }
        return dNodes;
    }
    exports.decorate = decorate;
    /**
     * Wrapper function for calls to create a widget.
     */
    function w(widgetConstructor, properties, children) {
        if (children === void 0) { children = []; }
        return {
            children: children,
            widgetConstructor: widgetConstructor,
            properties: properties,
            type: exports.WNODE
        };
    }
    exports.w = w;
    function v(tag, propertiesOrChildren, children) {
        if (propertiesOrChildren === void 0) { propertiesOrChildren = {}; }
        if (children === void 0) { children = []; }
        var properties = propertiesOrChildren;
        if (Array.isArray(propertiesOrChildren)) {
            children = propertiesOrChildren;
            properties = {};
        }
        var classes = properties.classes;
        if (typeof classes === 'function') {
            classes = classes();
            properties = lang_1.assign(properties, { classes: classes });
        }
        return {
            tag: tag,
            children: children,
            properties: properties,
            render: function () {
                return maquette_1.h(this.tag, this.properties, this.vNodes);
            },
            type: exports.HNODE
        };
    }
    exports.v = v;
});
