// @dojo/widget-core/diff.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    function isObjectOrArray(value) {
        return Object.prototype.toString.call(value) === '[object Object]' || Array.isArray(value);
    }
    function always(previousProperty, newProperty) {
        return {
            changed: true,
            value: newProperty
        };
    }
    function ignore(previousProperty, newProperty) {
        return {
            changed: false,
            value: newProperty
        };
    }
    function custom(previousProperty, newProperty, meta) {
        var diffFunction = meta.diffFunction, scope = meta.scope;
        if (!diffFunction) {
            return {
                changed: false,
                value: newProperty
            };
        }
        return diffFunction.call(scope, previousProperty, newProperty);
    }
    function reference(previousProperty, newProperty) {
        return {
            changed: previousProperty !== newProperty,
            value: newProperty
        };
    }
    function shallow(previousProperty, newProperty) {
        var changed = false;
        var validOldProperty = previousProperty && isObjectOrArray(previousProperty);
        var validNewProperty = newProperty && isObjectOrArray(newProperty);
        if (!validOldProperty || !validNewProperty) {
            return {
                changed: true,
                value: newProperty
            };
        }
        var previousKeys = Object.keys(previousProperty);
        var newKeys = Object.keys(newProperty);
        if (previousKeys.length !== newKeys.length) {
            changed = true;
        }
        else {
            changed = newKeys.some(function (key) {
                return newProperty[key] !== previousProperty[key];
            });
        }
        return {
            changed: changed,
            value: newProperty
        };
    }
    function diff(propertyName, diffDiffType, previousProperty, newProperty, meta) {
        var result;
        switch (diffDiffType) {
            case 1 /* CUSTOM */:
                result = custom(previousProperty, newProperty, meta);
                break;
            case 2 /* ALWAYS */:
                result = always(previousProperty, newProperty);
                break;
            case 3 /* IGNORE */:
                result = ignore(previousProperty, newProperty);
                break;
            case 4 /* REFERENCE */:
                result = reference(previousProperty, newProperty);
                break;
            case 5 /* SHALLOW */:
                result = shallow(previousProperty, newProperty);
                break;
            case 6 /* AUTO */:
                if (typeof newProperty === 'function') {
                    result = ignore(previousProperty, newProperty);
                }
                else if (isObjectOrArray(newProperty)) {
                    result = shallow(previousProperty, newProperty);
                }
                else {
                    result = reference(previousProperty, newProperty);
                }
                break;
            default:
                console.warn("no valid DiffType provided, will mark property '" + propertyName + "' as changed");
                result = always(previousProperty, newProperty);
        }
        return result;
    }
    exports.default = diff;
});
