// @dojo/widget-core/tsx.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "./d"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var d_1 = require("./d");
    exports.REGISTRY_ITEM = Symbol('Identifier for an item from the Widget Registry.');
    var FromRegistry = (function () {
        function FromRegistry() {
        }
        FromRegistry.type = exports.REGISTRY_ITEM;
        return FromRegistry;
    }());
    exports.FromRegistry = FromRegistry;
    function fromRegistry(tag) {
        return _a = (function (_super) {
                __extends(class_1, _super);
                function class_1() {
                    var _this = _super !== null && _super.apply(this, arguments) || this;
                    _this.name = tag;
                    return _this;
                }
                return class_1;
            }(FromRegistry)),
            _a.type = exports.REGISTRY_ITEM,
            _a;
        var _a;
    }
    exports.fromRegistry = fromRegistry;
    function spreadChildren(children, child) {
        if (Array.isArray(child)) {
            return child.reduce(spreadChildren, children);
        }
        else {
            return children.concat([child]);
        }
    }
    function tsx(tag, properties) {
        if (properties === void 0) { properties = {}; }
        var children = [];
        for (var _i = 2; _i < arguments.length; _i++) {
            children[_i - 2] = arguments[_i];
        }
        children = children.reduce(spreadChildren, []);
        properties = properties === null ? {} : properties;
        if (typeof tag === 'string') {
            return d_1.v(tag, properties, children);
        }
        else if (tag.type === exports.REGISTRY_ITEM) {
            var registryItem = new tag();
            return d_1.w(registryItem.name, properties, children);
        }
        else {
            return d_1.w(tag, properties, children);
        }
    }
    exports.tsx = tsx;
});
