// @dojo/widgets/slidepane/createSlidePaneElement.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "./SlidePane"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var SlidePane_1 = require("./SlidePane");
    /**
     * Configures a SlidePane web component
     */
    function createSlidePaneElement() {
        return {
            tagName: 'dojo-slide-pane',
            widgetConstructor: SlidePane_1.default,
            attributes: [
                {
                    attributeName: 'align'
                },
                {
                    attributeName: 'open',
                    value: function (value) { return value === 'false' || value === '0' ? false : true; }
                },
                {
                    attributeName: 'underlay',
                    value: function (value) { return value === 'false' || value === '0' ? false : true; }
                },
                {
                    attributeName: 'width',
                    value: function (value) { return Number(value); }
                }
            ],
            events: [
                {
                    propertyName: 'onOpen',
                    eventName: 'open'
                },
                {
                    propertyName: 'onRequestClose',
                    eventName: 'close'
                }
            ]
        };
    }
    exports.default = createSlidePaneElement;
    ;
});
