// @dojo/widgets/tabcontroller/Tab.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "tslib", "@dojo/widget-core/mixins/Themed", "@dojo/widget-core/d", "@dojo/widget-core/WidgetBase", "./styles/tabController.m.css"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var tslib_1 = require("tslib");
    var Themed_1 = require("@dojo/widget-core/mixins/Themed");
    var d_1 = require("@dojo/widget-core/d");
    var WidgetBase_1 = require("@dojo/widget-core/WidgetBase");
    var css = require("./styles/tabController.m.css");
    ;
    exports.TabBase = Themed_1.ThemedMixin(WidgetBase_1.WidgetBase);
    var Tab = /** @class */ (function (_super) {
        tslib_1.__extends(Tab, _super);
        function Tab() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        Tab.prototype.render = function () {
            var _a = this.properties, id = _a.id, labelledBy = _a.labelledBy;
            return d_1.v('div', {
                'aria-labelledby': labelledBy,
                classes: this.theme(css.tab),
                id: id,
                role: 'tabpanel'
            }, this.children);
        };
        Tab = tslib_1.__decorate([
            Themed_1.theme(css)
        ], Tab);
        return Tab;
    }(exports.TabBase));
    exports.default = Tab;
});
