// @dojo/widgets/tabcontroller/TabButton.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "tslib", "@dojo/widget-core/mixins/Themed", "@dojo/widget-core/d", "@dojo/widget-core/WidgetBase", "./styles/tabController.m.css"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var tslib_1 = require("tslib");
    var Themed_1 = require("@dojo/widget-core/mixins/Themed");
    var d_1 = require("@dojo/widget-core/d");
    var WidgetBase_1 = require("@dojo/widget-core/WidgetBase");
    var css = require("./styles/tabController.m.css");
    ;
    exports.TabButtonBase = Themed_1.ThemedMixin(WidgetBase_1.WidgetBase);
    var TabButton = /** @class */ (function (_super) {
        tslib_1.__extends(TabButton, _super);
        function TabButton() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        TabButton.prototype._onClick = function () {
            var _a = this.properties, disabled = _a.disabled, index = _a.index, onClick = _a.onClick;
            !disabled && onClick && onClick(index);
        };
        TabButton.prototype._onCloseClick = function (event) {
            var _a = this.properties, index = _a.index, onCloseClick = _a.onCloseClick;
            event.stopPropagation();
            onCloseClick && onCloseClick(index);
        };
        TabButton.prototype._onKeyDown = function (event) {
            var _a = this.properties, closeable = _a.closeable, disabled = _a.disabled, index = _a.index, onCloseClick = _a.onCloseClick, onDownArrowPress = _a.onDownArrowPress, onEndPress = _a.onEndPress, onHomePress = _a.onHomePress, onLeftArrowPress = _a.onLeftArrowPress, onRightArrowPress = _a.onRightArrowPress, onUpArrowPress = _a.onUpArrowPress;
            if (disabled) {
                return;
            }
            // Accessibility
            switch (event.which) {
                // Escape
                case 27 /* Escape */:
                    closeable && onCloseClick && onCloseClick(index);
                    break;
                // Left arrow
                case 37 /* Left */:
                    onLeftArrowPress && onLeftArrowPress();
                    break;
                // Right arrow
                case 39 /* Right */:
                    onRightArrowPress && onRightArrowPress();
                    break;
                // Up arrow
                case 38 /* Up */:
                    onUpArrowPress && onUpArrowPress();
                    break;
                // Down arrow
                case 40 /* Down */:
                    onDownArrowPress && onDownArrowPress();
                    break;
                // Home
                case 36 /* Home */:
                    onHomePress && onHomePress();
                    break;
                // End
                case 35 /* End */:
                    onEndPress && onEndPress();
                    break;
            }
        };
        TabButton.prototype._callFocus = function (element) {
            var _a = this.properties, callFocus = _a.callFocus, onFocusCalled = _a.onFocusCalled;
            if (callFocus) {
                element.focus();
                onFocusCalled && onFocusCalled();
            }
        };
        TabButton.prototype.getContent = function () {
            var _a = this.properties, active = _a.active, closeable = _a.closeable;
            return tslib_1.__spread(this.children, [
                closeable ? d_1.v('button', {
                    tabIndex: active ? 0 : -1,
                    classes: this.theme(css.close),
                    innerHTML: 'close tab',
                    onclick: this._onCloseClick
                }) : null
            ]);
        };
        TabButton.prototype.getModifierClasses = function () {
            var _a = this.properties, active = _a.active, disabled = _a.disabled;
            return [
                active ? css.activeTabButton : null,
                disabled ? css.disabledTabButton : null
            ];
        };
        TabButton.prototype.onElementCreated = function (element, key) {
            key === 'tab-button' && this._callFocus(element);
        };
        TabButton.prototype.onElementUpdated = function (element, key) {
            key === 'tab-button' && this._callFocus(element);
        };
        TabButton.prototype.render = function () {
            var _a = this.properties, active = _a.active, controls = _a.controls, disabled = _a.disabled, id = _a.id;
            return d_1.v('div', {
                'aria-controls': controls,
                'aria-disabled': disabled ? 'true' : 'false',
                'aria-selected': active ? 'true' : 'false',
                classes: this.theme(tslib_1.__spread([css.tabButton], this.getModifierClasses())),
                id: id,
                key: 'tab-button',
                onclick: this._onClick,
                onkeydown: this._onKeyDown,
                role: 'tab',
                tabIndex: active ? 0 : -1
            }, this.getContent());
        };
        TabButton = tslib_1.__decorate([
            Themed_1.theme(css)
        ], TabButton);
        return TabButton;
    }(exports.TabButtonBase));
    exports.default = TabButton;
});
