// @dojo/widgets/splitpane/createSplitPaneElement.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "./SplitPane"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var SplitPane_1 = require("./SplitPane");
    /**
     * Configures a SplitPane web component
     */
    function createSplitPaneElement() {
        return {
            tagName: 'dojo-split-pane',
            widgetConstructor: SplitPane_1.default,
            attributes: [
                {
                    attributeName: 'direction'
                },
                {
                    attributeName: 'size'
                }
            ],
            properties: [
                {
                    propertyName: 'leading'
                },
                {
                    propertyName: 'trailing'
                }
            ],
            events: [
                {
                    propertyName: 'onResize',
                    eventName: 'resize'
                }
            ]
        };
    }
    exports.default = createSplitPaneElement;
    ;
});
