// @dojo/widgets/splitpane/SplitPane.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "tslib", "@dojo/widget-core/mixins/Themed", "@dojo/widget-core/d", "@dojo/widget-core/WidgetBase", "./styles/splitPane.m.css"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var tslib_1 = require("tslib");
    var Themed_1 = require("@dojo/widget-core/mixins/Themed");
    var d_1 = require("@dojo/widget-core/d");
    var WidgetBase_1 = require("@dojo/widget-core/WidgetBase");
    var css = require("./styles/splitPane.m.css");
    ;
    ;
    exports.SplitPaneBase = Themed_1.ThemedMixin(WidgetBase_1.WidgetBase);
    var DEFAULT_SIZE = 100;
    var SplitPane = /** @class */ (function (_super) {
        tslib_1.__extends(SplitPane, _super);
        function SplitPane() {
            var _this = 
            /* istanbul ignore next: disregard transpiled `super`'s "else" block */
            _super.call(this) || this;
            /**
             * `mouseup` and other events aren't triggered when a user's cursor leaves div.root, so
             * global handlers are need to listen to the document instead. SplitPane
             * uses a `_dragging` flag so no handlers will be erroneously executed
             * if a user isn't actually resizing this SplitPane instance.
             */
            _this._boundHandlers = [];
            [
                { event: 'mouseup', func: _this._onDragEnd.bind(_this) },
                { event: 'mousemove', func: _this._onDragMove.bind(_this) },
                { event: 'touchmove', func: _this._onDragMove.bind(_this) }
            ].forEach(function (object) {
                document.addEventListener(object.event, object.func);
                _this._boundHandlers.push(object);
            });
            return _this;
        }
        SplitPane.prototype.onDetach = function () {
            this._boundHandlers.forEach(function (object) { return document.removeEventListener(object.event, object.func); });
        };
        SplitPane.prototype._deselect = function () {
            var selection = window.getSelection();
            selection.removeAllRanges();
        };
        SplitPane.prototype._getPosition = function (event) {
            var _a = this.properties.direction, direction = _a === void 0 ? 1 /* row */ : _a;
            if (direction === 1 /* row */) {
                return event.changedTouches ? event.changedTouches[0].clientX : event.clientX;
            }
            else {
                return event.changedTouches ? event.changedTouches[0].clientY : event.clientY;
            }
        };
        SplitPane.prototype._onDragStart = function (event) {
            this._dragging = true;
            this._position = this._getPosition(event);
            this._deselect();
        };
        SplitPane.prototype._onDragMove = function (event) {
            if (!this._dragging) {
                return;
            }
            this._deselect();
            var _a = this.properties, _b = _a.direction, direction = _b === void 0 ? 1 /* row */ : _b, onResize = _a.onResize, _c = _a.size, size = _c === void 0 ? DEFAULT_SIZE : _c;
            var currentPosition = this._getPosition(event);
            var newSize = (this._lastSize === undefined ? size : this._lastSize) + currentPosition - this._position;
            var maxSize = direction === 1 /* row */ ?
                this._root.offsetWidth - this._divider.offsetWidth :
                this._root.offsetHeight - this._divider.offsetHeight;
            this._lastSize = newSize;
            newSize = newSize < 0 ? 0 : newSize;
            newSize = newSize > maxSize ? maxSize : newSize;
            this._position = currentPosition;
            onResize && onResize(newSize);
        };
        SplitPane.prototype._onDragEnd = function (event) {
            this._dragging = false;
            this._lastSize = undefined;
        };
        SplitPane.prototype.getPaneContent = function (content) {
            return [content];
        };
        SplitPane.prototype.getPaneStyles = function () {
            var _a = this.properties, _b = _a.direction, direction = _b === void 0 ? 1 /* row */ : _b, _c = _a.size, size = _c === void 0 ? DEFAULT_SIZE : _c;
            var styles = {};
            styles[direction === 1 /* row */ ? 'width' : 'height'] = size + "px";
            return styles;
        };
        SplitPane.prototype.onElementCreated = function (element, key) {
            if (key === 'root') {
                this._root = element;
            }
            if (key === 'divider') {
                this._divider = element;
            }
        };
        SplitPane.prototype.render = function () {
            var _a = this.properties, _b = _a.direction, direction = _b === void 0 ? 1 /* row */ : _b, _c = _a.leading, leading = _c === void 0 ? null : _c, _d = _a.trailing, trailing = _d === void 0 ? null : _d;
            return d_1.v('div', {
                classes: tslib_1.__spread(this.theme([
                    css.root,
                    direction === 0 /* column */ ? css.column : css.row
                ]), [
                    css.rootFixed,
                    direction === 0 /* column */ ? css.columnFixed : css.rowFixed
                ]),
                key: 'root'
            }, [
                d_1.v('div', {
                    classes: [
                        this.theme(css.leading),
                        css.leadingFixed
                    ],
                    key: 'leading',
                    styles: this.getPaneStyles()
                }, this.getPaneContent(leading)),
                d_1.v('div', {
                    classes: [
                        this.theme(css.divider),
                        css.dividerFixed
                    ],
                    key: 'divider',
                    onmousedown: this._onDragStart,
                    ontouchend: this._onDragEnd,
                    ontouchstart: this._onDragStart
                }),
                d_1.v('div', {
                    classes: [
                        this.theme(css.trailing),
                        css.trailingFixed
                    ],
                    key: 'trailing'
                }, this.getPaneContent(trailing))
            ]);
        };
        SplitPane = tslib_1.__decorate([
            Themed_1.theme(css),
            tslib_1.__metadata("design:paramtypes", [])
        ], SplitPane);
        return SplitPane;
    }(exports.SplitPaneBase));
    exports.default = SplitPane;
});
