// @dojo/widgets/tooltip/createTooltipElement.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "./Tooltip"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var Tooltip_1 = require("./Tooltip");
    /**
     * Configures a Tooltip web component
     */
    function createTooltipElement() {
        return {
            tagName: 'dojo-tooltip',
            widgetConstructor: Tooltip_1.default,
            attributes: [
                {
                    attributeName: 'open',
                    value: function (value) { return value === 'false' || value === '0' ? false : true; }
                },
                {
                    attributeName: 'orientation'
                }
            ],
            properties: [
                {
                    propertyName: 'content'
                }
            ]
        };
    }
    exports.default = createTooltipElement;
    ;
});
