// @dojo/widgets/dialog/createDialogElement.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "./Dialog"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var Dialog_1 = require("./Dialog");
    /**
     * Configures a Dialog web component
     */
    function createDialogElement() {
        return {
            tagName: 'dojo-dialog',
            widgetConstructor: Dialog_1.default,
            attributes: [
                {
                    attributeName: 'closeable',
                    value: function (value) { return value === 'false' || value === '0' ? false : true; }
                },
                {
                    attributeName: 'closetext',
                    propertyName: 'closeText'
                },
                {
                    attributeName: 'enteranimation',
                    propertyName: 'enterAnimation'
                },
                {
                    attributeName: 'exitanimation',
                    propertyName: 'enterAnimation'
                },
                {
                    attributeName: 'modal',
                    value: function (value) { return value === 'false' || value === '0' ? false : true; }
                },
                {
                    attributeName: 'open',
                    value: function (value) { return value === 'false' || value === '0' ? false : true; }
                },
                {
                    attributeName: 'role'
                },
                {
                    attributeName: 'title'
                },
                {
                    attributeName: 'underlay',
                    value: function (value) { return value === 'false' || value === '0' ? false : true; }
                }
            ],
            events: [
                {
                    propertyName: 'onOpen',
                    eventName: 'open'
                },
                {
                    propertyName: 'onRequestClose',
                    eventName: 'close'
                }
            ]
        };
    }
    exports.default = createDialogElement;
    ;
});
