// @dojo/widgets/main.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "./button/Button", "./checkbox/Checkbox", "./combobox/ComboBox", "./dialog/Dialog", "./label/Label", "./radio/Radio", "./select/Select", "./slidepane/SlidePane", "./slider/Slider", "./splitpane/SplitPane", "./tabcontroller/Tab", "./tabcontroller/TabController", "./textarea/Textarea", "./textinput/TextInput", "./titlepane/TitlePane"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var Button_1 = require("./button/Button");
    exports.Button = Button_1.default;
    var Checkbox_1 = require("./checkbox/Checkbox");
    exports.Checkbox = Checkbox_1.default;
    var ComboBox_1 = require("./combobox/ComboBox");
    exports.ComboBox = ComboBox_1.default;
    var Dialog_1 = require("./dialog/Dialog");
    exports.Dialog = Dialog_1.default;
    var Label_1 = require("./label/Label");
    exports.Label = Label_1.default;
    var Radio_1 = require("./radio/Radio");
    exports.Radio = Radio_1.default;
    var Select_1 = require("./select/Select");
    exports.Select = Select_1.default;
    var SlidePane_1 = require("./slidepane/SlidePane");
    exports.SlidePane = SlidePane_1.default;
    var Slider_1 = require("./slider/Slider");
    exports.Slider = Slider_1.default;
    var SplitPane_1 = require("./splitpane/SplitPane");
    exports.SplitPane = SplitPane_1.default;
    var Tab_1 = require("./tabcontroller/Tab");
    exports.Tab = Tab_1.default;
    var TabController_1 = require("./tabcontroller/TabController");
    exports.TabController = TabController_1.default;
    var Textarea_1 = require("./textarea/Textarea");
    exports.Textarea = Textarea_1.default;
    var TextInput_1 = require("./textinput/TextInput");
    exports.TextInput = TextInput_1.default;
    var TitlePane_1 = require("./titlepane/TitlePane");
    exports.TitlePane = TitlePane_1.default;
});
