// @dojo/widgets/titlepane/createTitlePaneElement.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "./TitlePane"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var TitlePane_1 = require("./TitlePane");
    /**
     * Configures a TitlePane web component
     */
    function createTitlePaneElement() {
        return {
            tagName: 'dojo-title-pane',
            widgetConstructor: TitlePane_1.default,
            attributes: [
                {
                    attributeName: 'headinglevel',
                    value: function (value) { return Number(value); }
                },
                {
                    attributeName: 'closeable',
                    value: function (value) { return value === 'false' || value === '0' ? false : true; }
                },
                {
                    attributeName: 'open',
                    value: function (value) { return value === 'false' || value === '0' ? false : true; }
                },
                {
                    attributeName: 'title'
                }
            ],
            events: [
                {
                    propertyName: 'onRequestClose',
                    eventName: 'requestClose'
                },
                {
                    propertyName: 'onRequestOpen',
                    eventName: 'requestOpen'
                }
            ]
        };
    }
    exports.default = createTitlePaneElement;
});
