// @dojo/widgets/titlepane/TitlePane.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "@dojo/core/uuid", "@dojo/widget-core/mixins/Themeable", "@dojo/widget-core/d", "@dojo/widget-core/WidgetBase", "./styles/titlePane.m.css", "../common/styles/icons.m.css"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var uuid_1 = require("@dojo/core/uuid");
    var Themeable_1 = require("@dojo/widget-core/mixins/Themeable");
    var d_1 = require("@dojo/widget-core/d");
    var WidgetBase_1 = require("@dojo/widget-core/WidgetBase");
    var css = require("./styles/titlePane.m.css");
    var iconCss = require("../common/styles/icons.m.css");
    ;
    exports.TitlePaneBase = Themeable_1.ThemeableMixin(WidgetBase_1.WidgetBase);
    var TitlePane = /** @class */ (function (_super) {
        __extends(TitlePane, _super);
        function TitlePane() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this._contentId = uuid_1.default();
            _this._titleId = uuid_1.default();
            return _this;
        }
        TitlePane.prototype._afterRender = function (element) {
            var _this = this;
            // Conditionally adjust top margin. Done manually instead of through Maquette
            // so the underlying DOM is accessible, as we need to know the content height.
            // Put in a rAF to push this operation to the next tick, otherwise
            // element.offsetHeight can be incorrect (e.g. before styling is applied)
            // Note that this will go away when meta support is added to widget-core
            requestAnimationFrame(function () {
                var _a = _this.properties.open, open = _a === void 0 ? true : _a;
                var height = element.offsetHeight;
                element.style.marginTop = open ? '0px' : "-" + height + "px";
            });
        };
        TitlePane.prototype._onTitleClick = function () {
            this._toggle();
        };
        TitlePane.prototype._onTitleKeyUp = function (event) {
            var keyCode = event.keyCode;
            if (keyCode === 13 /* Enter */ || keyCode === 32 /* Space */) {
                this._toggle();
            }
        };
        TitlePane.prototype._toggle = function () {
            var _a = this.properties, _b = _a.closeable, closeable = _b === void 0 ? true : _b, onRequestClose = _a.onRequestClose, onRequestOpen = _a.onRequestOpen, _c = _a.open, open = _c === void 0 ? true : _c;
            if (!closeable) {
                return;
            }
            if (open) {
                onRequestClose && onRequestClose(this);
            }
            else {
                onRequestOpen && onRequestOpen(this);
            }
        };
        TitlePane.prototype.onElementCreated = function (element, key) {
            key === 'content' && this._afterRender(element);
        };
        TitlePane.prototype.onElementUpdated = function (element, key) {
            key === 'content' && this._afterRender(element);
        };
        TitlePane.prototype.render = function () {
            var _a = this.properties, _b = _a.closeable, closeable = _b === void 0 ? true : _b, headingLevel = _a.headingLevel, _c = _a.open, open = _c === void 0 ? true : _c, title = _a.title;
            return d_1.v('div', {
                classes: this.classes(css.root, open ? css.open : null).fixed(css.rootFixed)
            }, [
                d_1.v('div', {
                    'aria-level': headingLevel ? String(headingLevel) : null,
                    classes: this.classes(closeable ? css.closeable : null, css.title).fixed(closeable ? css.closeableFixed : null, css.titleFixed),
                    onclick: this._onTitleClick,
                    onkeyup: this._onTitleKeyUp,
                    role: 'heading'
                }, [
                    d_1.v('div', {
                        'aria-controls': this._contentId,
                        'aria-disabled': closeable ? null : 'true',
                        'aria-expanded': String(open),
                        classes: this.classes(css.titleButton),
                        id: this._titleId,
                        role: 'button',
                        tabIndex: closeable ? 0 : -1
                    }, [
                        d_1.v('i', {
                            classes: this.classes(css.arrow, iconCss.icon, open ? iconCss.downIcon : iconCss.rightIcon),
                            role: 'presentation',
                            'aria-hidden': 'true'
                        }),
                        title
                    ])
                ]),
                d_1.v('div', {
                    'aria-hidden': open ? null : 'true',
                    'aria-labelledby': this._titleId,
                    classes: this.classes(css.content),
                    id: this._contentId,
                    key: 'content'
                }, this.children)
            ]);
        };
        TitlePane = __decorate([
            Themeable_1.theme(css),
            Themeable_1.theme(iconCss)
        ], TitlePane);
        return TitlePane;
    }(exports.TitlePaneBase));
    exports.default = TitlePane;
});
