// @dojo/widgets/accordionpane/createAccordionPaneElement.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "./AccordionPane"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var AccordionPane_1 = require("./AccordionPane");
    /**
     * Configures a AccordionPane web component
     */
    function createAccordionPaneElement() {
        return {
            tagName: 'dojo-accordionpane',
            widgetConstructor: AccordionPane_1.default,
            properties: [
                {
                    propertyName: 'openKeys'
                }
            ],
            events: [
                {
                    propertyName: 'onRequestClose',
                    eventName: 'requestClose'
                },
                {
                    propertyName: 'onRequestOpen',
                    eventName: 'requestOpen'
                }
            ]
        };
    }
    exports.default = createAccordionPaneElement;
    ;
});
