// @dojo/shim/Set.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "./global", "./iterator", "./support/has", "./Symbol"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var global_1 = require("./global");
    var iterator_1 = require("./iterator");
    var has_1 = require("./support/has");
    require("./Symbol");
    exports.Set = global_1.default.Set;
    if (!has_1.default('es6-set')) {
        exports.Set = (_a = (function () {
                function Set(iterable) {
                    var _this = this;
                    this._setData = [];
                    this[Symbol.toStringTag] = 'Set';
                    if (iterable) {
                        iterator_1.forOf(iterable, function (value) { return _this.add(value); });
                    }
                }
                ;
                Set.prototype.add = function (value) {
                    if (this.has(value)) {
                        return this;
                    }
                    this._setData.push(value);
                    return this;
                };
                ;
                Set.prototype.clear = function () {
                    this._setData.length = 0;
                };
                ;
                Set.prototype.delete = function (value) {
                    var idx = this._setData.indexOf(value);
                    if (idx === -1) {
                        return false;
                    }
                    this._setData.splice(idx, 1);
                    return true;
                };
                ;
                Set.prototype.entries = function () {
                    return new iterator_1.ShimIterator(this._setData.map(function (value) { return [value, value]; }));
                };
                ;
                Set.prototype.forEach = function (callbackfn, thisArg) {
                    var iterator = this.values();
                    var result = iterator.next();
                    while (!result.done) {
                        callbackfn.call(thisArg, result.value, result.value, this);
                        result = iterator.next();
                    }
                };
                ;
                Set.prototype.has = function (value) {
                    return this._setData.indexOf(value) > -1;
                };
                ;
                Set.prototype.keys = function () {
                    return new iterator_1.ShimIterator(this._setData);
                };
                ;
                Object.defineProperty(Set.prototype, "size", {
                    get: function () {
                        return this._setData.length;
                    },
                    enumerable: true,
                    configurable: true
                });
                ;
                Set.prototype.values = function () {
                    return new iterator_1.ShimIterator(this._setData);
                };
                ;
                Set.prototype[Symbol.iterator] = function () {
                    return new iterator_1.ShimIterator(this._setData);
                };
                ;
                return Set;
            }()),
            _a[Symbol.species] = _a,
            _a);
    }
    exports.default = exports.Set;
    var _a;
});
