// @dojo/widgets/label/Label.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "@dojo/widget-core/WidgetBase", "@dojo/widget-core/mixins/Themeable", "@dojo/widget-core/mixins/Registry", "@dojo/widget-core/d", "@dojo/core/lang", "../common/styles/base.m.css"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var WidgetBase_1 = require("@dojo/widget-core/WidgetBase");
    var Themeable_1 = require("@dojo/widget-core/mixins/Themeable");
    var Registry_1 = require("@dojo/widget-core/mixins/Registry");
    var d_1 = require("@dojo/widget-core/d");
    var lang_1 = require("@dojo/core/lang");
    var baseCss = require("../common/styles/base.m.css");
    /**
     * Default settings for form labels
     */
    var labelDefaults = {
        content: '',
        before: true,
        hidden: false
    };
    exports.LabelBase = Registry_1.RegistryMixin(Themeable_1.ThemeableMixin(WidgetBase_1.WidgetBase));
    var Label = (function (_super) {
        __extends(Label, _super);
        function Label() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        Label.prototype.render = function () {
            var _a = this.properties, formId = _a.formId, label = _a.label, _b = _a.classes, classes = _b === void 0 ? {} : _b;
            // assign string or object label properites with defaults
            var labelProps;
            if (typeof label === 'string') {
                labelProps = lang_1.assign({}, labelDefaults, { content: label });
            }
            else {
                labelProps = lang_1.assign({}, labelDefaults, label);
            }
            // add label text node to children
            var labelText = d_1.v('span', {
                innerHTML: labelProps.content,
                classes: this.classes().fixed(labelProps.hidden ? baseCss.visuallyHidden : null)
            });
            if (labelProps.before) {
                this.children.unshift(labelText);
            }
            else {
                this.children.push(labelText);
            }
            return d_1.v('label', {
                'form': formId,
                classes: classes
            }, this.children);
        };
        return Label;
    }(exports.LabelBase));
    Label = __decorate([
        Themeable_1.theme(baseCss)
    ], Label);
    exports.default = Label;
});
