// @dojo/widgets/calendar/CalendarCell.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "tslib", "@dojo/widget-core/WidgetBase", "@dojo/widget-core/mixins/Themed", "@dojo/widget-core/d", "./styles/calendar.m.css"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var tslib_1 = require("tslib");
    var WidgetBase_1 = require("@dojo/widget-core/WidgetBase");
    var Themed_1 = require("@dojo/widget-core/mixins/Themed");
    var d_1 = require("@dojo/widget-core/d");
    var css = require("./styles/calendar.m.css");
    ;
    exports.CalendarCellBase = Themed_1.ThemedMixin(WidgetBase_1.WidgetBase);
    var CalendarCell = /** @class */ (function (_super) {
        tslib_1.__extends(CalendarCell, _super);
        function CalendarCell() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        CalendarCell.prototype.onElementCreated = function (element, key) {
            this._callFocus(element);
        };
        CalendarCell.prototype.onElementUpdated = function (element, key) {
            this._callFocus(element);
        };
        CalendarCell.prototype._callFocus = function (element) {
            var _a = this.properties, callFocus = _a.callFocus, onFocusCalled = _a.onFocusCalled;
            if (callFocus) {
                element.focus();
                onFocusCalled && onFocusCalled();
            }
        };
        CalendarCell.prototype._onClick = function (event) {
            var _a = this.properties, date = _a.date, _b = _a.disabled, disabled = _b === void 0 ? false : _b, onClick = _a.onClick;
            onClick && onClick(date, disabled);
        };
        CalendarCell.prototype._onKeyDown = function (event) {
            var onKeyDown = this.properties.onKeyDown;
            onKeyDown && onKeyDown(event);
        };
        CalendarCell.prototype.formatDate = function (date) {
            return d_1.v('span', ["" + date]);
        };
        CalendarCell.prototype.getModifierClasses = function () {
            var _a = this.properties, _b = _a.disabled, disabled = _b === void 0 ? false : _b, _c = _a.selected, selected = _c === void 0 ? false : _c, _d = _a.today, today = _d === void 0 ? false : _d;
            return [
                disabled ? css.inactiveDate : null,
                selected ? css.selectedDate : null,
                today ? css.todayDate : null
            ];
        };
        CalendarCell.prototype.render = function () {
            var _a = this.properties, date = _a.date, _b = _a.focusable, focusable = _b === void 0 ? false : _b, _c = _a.selected, selected = _c === void 0 ? false : _c;
            return d_1.v('td', {
                key: 'root',
                role: 'gridcell',
                'aria-selected': "" + selected,
                tabIndex: focusable ? 0 : -1,
                classes: this.theme(tslib_1.__spread([css.date], this.getModifierClasses())),
                onclick: this._onClick,
                onkeydown: this._onKeyDown
            }, [this.formatDate(date)]);
        };
        CalendarCell = tslib_1.__decorate([
            Themed_1.theme(css)
        ], CalendarCell);
        return CalendarCell;
    }(exports.CalendarCellBase));
    exports.default = CalendarCell;
});
