// @dojo/widgets/calendar/Calendar.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "tslib", "@dojo/widget-core/WidgetBase", "@dojo/widget-core/mixins/Themed", "@dojo/widget-core/d", "@dojo/core/uuid", "./DatePicker", "./CalendarCell", "./styles/calendar.m.css", "../common/styles/base.m.css", "../common/styles/icons.m.css"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var tslib_1 = require("tslib");
    var WidgetBase_1 = require("@dojo/widget-core/WidgetBase");
    var Themed_1 = require("@dojo/widget-core/mixins/Themed");
    var d_1 = require("@dojo/widget-core/d");
    var uuid_1 = require("@dojo/core/uuid");
    var DatePicker_1 = require("./DatePicker");
    var CalendarCell_1 = require("./CalendarCell");
    var css = require("./styles/calendar.m.css");
    var baseCss = require("../common/styles/base.m.css");
    var iconCss = require("../common/styles/icons.m.css");
    exports.DEFAULT_MONTHS = [
        { short: 'Jan', long: 'January' },
        { short: 'Feb', long: 'February' },
        { short: 'Mar', long: 'March' },
        { short: 'Apr', long: 'April' },
        { short: 'May', long: 'May' },
        { short: 'Jun', long: 'June' },
        { short: 'Jul', long: 'July' },
        { short: 'Aug', long: 'August' },
        { short: 'Sep', long: 'September' },
        { short: 'Oct', long: 'October' },
        { short: 'Nov', long: 'November' },
        { short: 'Dec', long: 'December' }
    ];
    exports.DEFAULT_WEEKDAYS = [
        { short: 'Sun', long: 'Sunday' },
        { short: 'Mon', long: 'Monday' },
        { short: 'Tue', long: 'Tuesday' },
        { short: 'Wed', long: 'Wednesday' },
        { short: 'Thu', long: 'Thursday' },
        { short: 'Fri', long: 'Friday' },
        { short: 'Sat', long: 'Saturday' }
    ];
    exports.DEFAULT_LABELS = {
        chooseMonth: 'Choose Month',
        chooseYear: 'Choose Year',
        previousMonth: 'Previous Month',
        nextMonth: 'Next Month'
    };
    exports.CalendarBase = Themed_1.ThemedMixin(WidgetBase_1.WidgetBase);
    var Calendar = /** @class */ (function (_super) {
        tslib_1.__extends(Calendar, _super);
        function Calendar() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this._callDateFocus = false;
            _this._defaultDate = new Date();
            _this._focusedDay = 1;
            _this._monthLabelId = uuid_1.default();
            _this._popupOpen = false;
            return _this;
        }
        Calendar.prototype._getMonthLength = function (month, year) {
            var lastDate = new Date(year, month + 1, 0);
            return lastDate.getDate();
        };
        Calendar.prototype._getMonthYear = function () {
            var _a = this.properties, month = _a.month, _b = _a.selectedDate, selectedDate = _b === void 0 ? this._defaultDate : _b, year = _a.year;
            return {
                month: typeof month === 'number' ? month : selectedDate.getMonth(),
                year: typeof year === 'number' ? year : selectedDate.getFullYear()
            };
        };
        Calendar.prototype._goToDate = function (day) {
            var _a = this._getMonthYear(), month = _a.month, year = _a.year;
            var currentMonthLength = this._getMonthLength(month, year);
            var previousMonthLength = this._getMonthLength(month - 1, year);
            if (day < 1) {
                this._onMonthDecrement();
                day += previousMonthLength;
            }
            else if (day > currentMonthLength) {
                this._onMonthIncrement();
                day -= currentMonthLength;
            }
            this._focusedDay = day;
            this._callDateFocus = true;
            this.invalidate();
        };
        Calendar.prototype._onDateClick = function (date, disabled) {
            var onDateSelect = this.properties.onDateSelect;
            var _a = this._getMonthYear(), month = _a.month, year = _a.year;
            if (disabled && date < 15) {
                (_b = this._onMonthIncrement(), month = _b.month, year = _b.year);
                this._callDateFocus = true;
            }
            else if (disabled && date >= 15) {
                (_c = this._onMonthDecrement(), month = _c.month, year = _c.year);
                this._callDateFocus = true;
            }
            this._focusedDay = date;
            onDateSelect && onDateSelect(new Date(year, month, date));
            var _b, _c;
        };
        Calendar.prototype._onDateFocusCalled = function () {
            this._callDateFocus = false;
        };
        Calendar.prototype._onDateKeyDown = function (event) {
            var _a = this._getMonthYear(), month = _a.month, year = _a.year;
            switch (event.which) {
                case 38 /* Up */:
                    event.preventDefault();
                    this._goToDate(this._focusedDay - 7);
                    break;
                case 40 /* Down */:
                    event.preventDefault();
                    this._goToDate(this._focusedDay + 7);
                    break;
                case 37 /* Left */:
                    event.preventDefault();
                    this._goToDate(this._focusedDay - 1);
                    break;
                case 39 /* Right */:
                    event.preventDefault();
                    this._goToDate(this._focusedDay + 1);
                    break;
                case 33 /* PageUp */:
                    event.preventDefault();
                    this._goToDate(1);
                    break;
                case 34 /* PageDown */:
                    event.preventDefault();
                    var monthLengh = this._getMonthLength(month, year);
                    this._goToDate(monthLengh);
                    break;
                case 13 /* Enter */:
                case 32 /* Space */:
                    var onDateSelect = this.properties.onDateSelect;
                    onDateSelect && onDateSelect(new Date(year, month, this._focusedDay));
            }
        };
        Calendar.prototype._onMonthDecrement = function () {
            var _a = this._getMonthYear(), month = _a.month, year = _a.year;
            var _b = this.properties, onMonthChange = _b.onMonthChange, onYearChange = _b.onYearChange;
            if (month === 0) {
                onMonthChange && onMonthChange(11);
                onYearChange && onYearChange(year - 1);
                return { month: 11, year: year - 1 };
            }
            onMonthChange && onMonthChange(month - 1);
            return { month: month - 1, year: year };
        };
        Calendar.prototype._onMonthIncrement = function () {
            var _a = this._getMonthYear(), month = _a.month, year = _a.year;
            var _b = this.properties, onMonthChange = _b.onMonthChange, onYearChange = _b.onYearChange;
            if (month === 11) {
                onMonthChange && onMonthChange(0);
                onYearChange && onYearChange(year + 1);
                return { month: 0, year: year + 1 };
            }
            onMonthChange && onMonthChange(month + 1);
            return { month: month + 1, year: year };
        };
        Calendar.prototype._onMonthPageDown = function () {
            this._onMonthDecrement();
        };
        Calendar.prototype._onMonthPageUp = function () {
            this._onMonthIncrement();
        };
        Calendar.prototype._renderDateGrid = function (selectedDate) {
            var _a = this._getMonthYear(), month = _a.month, year = _a.year;
            var currentMonthLength = this._getMonthLength(month, year);
            var previousMonthLength = this._getMonthLength(month - 1, year);
            var initialWeekday = new Date(year, month, 1).getDay();
            var todayString = new Date().toDateString();
            var dayIndex = 0;
            var date = initialWeekday > 0 ? previousMonthLength - initialWeekday : 0;
            var isCurrentMonth = initialWeekday > 0 ? false : true;
            var isSelectedDay;
            var weeks = [];
            var days;
            var dateString;
            var i;
            for (var week = 0; week < 6; week++) {
                days = [];
                for (i = 0; i < 7; i++) {
                    // find the next date
                    // if we've reached the end of the previous month, reset to 1
                    if (date > dayIndex && date >= previousMonthLength) {
                        date = 1;
                        isCurrentMonth = true;
                    }
                    else if (date <= dayIndex && date >= currentMonthLength) {
                        date = 1;
                        isCurrentMonth = false;
                    }
                    else {
                        date++;
                    }
                    dayIndex++;
                    // set isSelectedDay if the dates match
                    dateString = new Date(year, month, date).toDateString();
                    if (isCurrentMonth && selectedDate && dateString === selectedDate.toDateString()) {
                        isSelectedDay = true;
                    }
                    else {
                        isSelectedDay = false;
                    }
                    var isToday = isCurrentMonth && dateString === todayString;
                    days.push(this.renderDateCell(date, week * 7 + i, isSelectedDay, isCurrentMonth, isToday));
                }
                weeks.push(d_1.v('tr', days));
            }
            return weeks;
        };
        Calendar.prototype.renderDateCell = function (date, index, selected, currentMonth, today) {
            var _a = this.properties.theme, theme = _a === void 0 ? {} : _a;
            return d_1.w(CalendarCell_1.default, {
                key: "date-" + index,
                callFocus: this._callDateFocus && currentMonth && date === this._focusedDay,
                date: date,
                disabled: !currentMonth,
                focusable: currentMonth && date === this._focusedDay,
                selected: selected,
                theme: theme,
                today: today,
                onClick: this._onDateClick,
                onFocusCalled: this._onDateFocusCalled,
                onKeyDown: this._onDateKeyDown
            });
        };
        Calendar.prototype.renderDatePicker = function () {
            var _this = this;
            var _a = this.properties, _b = _a.labels, labels = _b === void 0 ? exports.DEFAULT_LABELS : _b, _c = _a.monthNames, monthNames = _c === void 0 ? exports.DEFAULT_MONTHS : _c, renderMonthLabel = _a.renderMonthLabel, _d = _a.theme, theme = _d === void 0 ? {} : _d, onMonthChange = _a.onMonthChange, onYearChange = _a.onYearChange;
            var _e = this._getMonthYear(), month = _e.month, year = _e.year;
            return d_1.w(DatePicker_1.default, {
                key: 'date-picker',
                labelId: this._monthLabelId,
                labels: labels,
                month: month,
                monthNames: monthNames,
                renderMonthLabel: renderMonthLabel,
                theme: theme,
                year: year,
                onPopupChange: function (open) {
                    _this._popupOpen = open;
                },
                onRequestMonthChange: function (requestMonth) {
                    onMonthChange && onMonthChange(requestMonth);
                },
                onRequestYearChange: function (requestYear) {
                    onYearChange && onYearChange(requestYear);
                }
            });
        };
        Calendar.prototype.renderPagingButtonContent = function (type) {
            var _a = this.properties.labels, labels = _a === void 0 ? exports.DEFAULT_LABELS : _a;
            var iconClass = type === "next" /* next */ ? iconCss.rightIcon : iconCss.leftIcon;
            var labelText = type === "next" /* next */ ? labels.nextMonth : labels.previousMonth;
            return [
                d_1.v('i', { classes: this.theme([iconCss.icon, iconClass]),
                    role: 'presentation', 'aria-hidden': 'true'
                }),
                d_1.v('span', { classes: [baseCss.visuallyHidden] }, [labelText])
            ];
        };
        Calendar.prototype.renderWeekdayCell = function (day) {
            var renderWeekdayCell = this.properties.renderWeekdayCell;
            return renderWeekdayCell ? renderWeekdayCell(day) : d_1.v('abbr', { title: day.long }, [day.short]);
        };
        Calendar.prototype.render = function () {
            var _a = this.properties, selectedDate = _a.selectedDate, _b = _a.weekdayNames, weekdayNames = _b === void 0 ? exports.DEFAULT_WEEKDAYS : _b;
            // Calendar Weekday array
            var weekdays = [];
            for (var weekday in weekdayNames) {
                weekdays.push(d_1.v('th', {
                    role: 'columnheader',
                    classes: this.theme(css.weekday)
                }, [
                    this.renderWeekdayCell(weekdayNames[weekday])
                ]));
            }
            return d_1.v('div', { classes: this.theme(css.root) }, [
                // header
                this.renderDatePicker(),
                // date table
                d_1.v('table', {
                    cellspacing: '0',
                    cellpadding: '0',
                    role: 'grid',
                    'aria-labelledby': this._monthLabelId,
                    classes: [this.theme(css.dateGrid), this._popupOpen ? baseCss.visuallyHidden : null]
                }, [
                    d_1.v('thead', [
                        d_1.v('tr', weekdays)
                    ]),
                    d_1.v('tbody', this._renderDateGrid(selectedDate))
                ]),
                // controls
                d_1.v('div', {
                    classes: [this.theme(css.controls), this._popupOpen ? baseCss.visuallyHidden : null]
                }, [
                    d_1.v('button', {
                        classes: this.theme(css.previous),
                        tabIndex: this._popupOpen ? -1 : 0,
                        onclick: this._onMonthPageDown
                    }, this.renderPagingButtonContent("previous" /* previous */)),
                    d_1.v('button', {
                        classes: this.theme(css.next),
                        tabIndex: this._popupOpen ? -1 : 0,
                        onclick: this._onMonthPageUp
                    }, this.renderPagingButtonContent("next" /* next */))
                ])
            ]);
        };
        Calendar = tslib_1.__decorate([
            Themed_1.theme(css),
            Themed_1.theme(iconCss)
        ], Calendar);
        return Calendar;
    }(exports.CalendarBase));
    exports.default = Calendar;
});
