// @dojo/widgets/calendar/createCalendarElement.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "./Calendar"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var Calendar_1 = require("./Calendar");
    /**
     * Configures a Calendar web component
     */
    function createCalendarElement() {
        return {
            tagName: 'dojo-calendar',
            widgetConstructor: Calendar_1.default,
            attributes: [
                {
                    attributeName: 'month',
                    value: function (value) { return value ? parseInt(value, 10) : undefined; }
                },
                {
                    attributeName: 'selecteddate',
                    propertyName: 'selectedDate',
                    value: function (value) { return value ? new Date(value) : undefined; }
                },
                {
                    attributeName: 'year',
                    value: function (value) { return value ? parseInt(value, 10) : undefined; }
                }
            ],
            properties: [
                { propertyName: 'labels' },
                { propertyName: 'monthNames' },
                { propertyName: 'weekdayNames' },
                { propertyName: 'renderMonthLabel' },
                { propertyName: 'renderWeekdayCell' }
            ],
            events: [
                {
                    propertyName: 'onMonthChange',
                    eventName: 'monthchange'
                },
                {
                    propertyName: 'onYearChange',
                    eventName: 'yearchange'
                },
                {
                    propertyName: 'onDateSelect',
                    eventName: 'dateselect'
                }
            ]
        };
    }
    exports.default = createCalendarElement;
    ;
});
