// @dojo/widgets/calendar/DatePicker.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "tslib", "@dojo/widget-core/WidgetBase", "@dojo/widget-core/mixins/Themed", "@dojo/widget-core/d", "@dojo/core/uuid", "./styles/calendar.m.css", "../common/styles/base.m.css", "../common/styles/icons.m.css"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var tslib_1 = require("tslib");
    var WidgetBase_1 = require("@dojo/widget-core/WidgetBase");
    var Themed_1 = require("@dojo/widget-core/mixins/Themed");
    var d_1 = require("@dojo/widget-core/d");
    var uuid_1 = require("@dojo/core/uuid");
    var css = require("./styles/calendar.m.css");
    var baseCss = require("../common/styles/base.m.css");
    var iconCss = require("../common/styles/icons.m.css");
    ;
    ;
    ;
    ;
    exports.DatePickerBase = Themed_1.ThemedMixin(WidgetBase_1.WidgetBase);
    var BASE_YEAR = 2000;
    var DatePicker = /** @class */ (function (_super) {
        tslib_1.__extends(DatePicker, _super);
        function DatePicker() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this._callMonthTriggerFocus = false;
            _this._callYearTriggerFocus = false;
            _this._callMonthPopupFocus = false;
            _this._callYearPopupFocus = false;
            _this._idBase = uuid_1.default();
            _this._monthPopupOpen = false;
            _this._yearPopupOpen = false;
            _this._yearPage = 0;
            return _this;
        }
        DatePicker.prototype._closeMonthPopup = function () {
            var onPopupChange = this.properties.onPopupChange;
            this._monthPopupOpen = false;
            this._callMonthTriggerFocus = true;
            this.invalidate();
            onPopupChange && onPopupChange(this._getPopupState());
        };
        DatePicker.prototype._closeYearPopup = function () {
            var onPopupChange = this.properties.onPopupChange;
            this._yearPopupOpen = false;
            this._callYearTriggerFocus = true;
            this.invalidate();
            onPopupChange && onPopupChange(this._getPopupState());
        };
        DatePicker.prototype._getPopupState = function () {
            return this._monthPopupOpen || this._yearPopupOpen;
        };
        DatePicker.prototype._getYearRange = function () {
            var _a = this.properties, year = _a.year, _b = _a.yearRange, yearRange = _b === void 0 ? 20 : _b;
            var offset = (year - BASE_YEAR) % yearRange - yearRange * this._yearPage;
            if (year >= BASE_YEAR) {
                return { first: year - offset, last: year + yearRange - offset };
            }
            else {
                return { first: year - (yearRange + offset), last: year - offset };
            }
        };
        // move focus when opening/closing the popup
        DatePicker.prototype.onElementUpdated = function (element, key) {
            if (key === 'month-button') {
                if (!this._monthPopupOpen && this._callMonthTriggerFocus) {
                    element.focus();
                    this._callMonthTriggerFocus = false;
                }
            }
            if (key === 'year-button') {
                if (!this._yearPopupOpen && this._callYearTriggerFocus) {
                    element.focus();
                    this._callYearTriggerFocus = false;
                }
            }
            if (this._callMonthPopupFocus && key.indexOf(this._idBase + "_month_radios") > -1) {
                var month = key.split('_')[3];
                if (this._monthPopupOpen && month === "" + this.properties.month) {
                    element.children[0].focus();
                    this._callMonthPopupFocus = false;
                }
            }
            if (this._callYearPopupFocus && key.indexOf(this._idBase + "_year_radios") > -1) {
                var year = key.split('_')[3];
                if (this._yearPopupOpen && year === "" + this.properties.year) {
                    element.children[0].focus();
                    this._callYearPopupFocus = false;
                }
            }
        };
        DatePicker.prototype._onMonthButtonClick = function () {
            this._monthPopupOpen ? this._closeMonthPopup() : this._openMonthPopup();
        };
        DatePicker.prototype._onMonthRadioChange = function (event) {
            var onRequestMonthChange = this.properties.onRequestMonthChange;
            onRequestMonthChange && onRequestMonthChange(parseInt(event.target.value, 10));
        };
        DatePicker.prototype._onPopupKeyDown = function (event) {
            // close popup on escape, or if a value is selected with enter/space
            if (event.which === 27 /* Escape */ ||
                event.which === 13 /* Enter */ ||
                event.which === 32 /* Space */) {
                this._closeMonthPopup();
                this._closeYearPopup();
            }
        };
        DatePicker.prototype._onYearButtonClick = function () {
            this._yearPopupOpen ? this._closeYearPopup() : this._openYearPopup();
        };
        DatePicker.prototype._onYearPageDown = function () {
            this._yearPage--;
            this._yearPopupOpen && this.invalidate();
        };
        DatePicker.prototype._onYearPageUp = function () {
            this._yearPage++;
            this._yearPopupOpen && this.invalidate();
        };
        DatePicker.prototype._onYearRadioChange = function (event) {
            var onRequestYearChange = this.properties.onRequestYearChange;
            this._yearPage = 0;
            onRequestYearChange && onRequestYearChange(parseInt(event.target.value, 10));
        };
        DatePicker.prototype._openMonthPopup = function () {
            var onPopupChange = this.properties.onPopupChange;
            this._monthPopupOpen = true;
            this._callMonthPopupFocus = true;
            this._yearPopupOpen = false;
            this.invalidate();
            onPopupChange && onPopupChange(this._getPopupState());
        };
        DatePicker.prototype._openYearPopup = function () {
            var onPopupChange = this.properties.onPopupChange;
            this._yearPopupOpen = true;
            this._callYearPopupFocus = true;
            this._monthPopupOpen = false;
            this.invalidate();
            onPopupChange && onPopupChange(this._getPopupState());
        };
        DatePicker.prototype.renderControlsTrigger = function (type) {
            var _a = this.properties, month = _a.month, monthNames = _a.monthNames, year = _a.year;
            var content = type === "month" /* month */ ? monthNames[month].long : "" + year;
            var open = type === "month" /* month */ ? this._monthPopupOpen : this._yearPopupOpen;
            var onclick = type === "month" /* month */ ? this._onMonthButtonClick : this._onYearButtonClick;
            return d_1.v('button', {
                key: type + "-button",
                'aria-controls': this._idBase + "_" + type + "_dialog",
                'aria-expanded': "" + open,
                'aria-haspopup': 'true',
                id: this._idBase + "_" + type + "_button",
                classes: this.theme([
                    css[type + "Trigger"],
                    open ? css[type + "TriggerActive"] : null
                ]),
                role: 'menuitem',
                onclick: onclick
            }, [content]);
        };
        DatePicker.prototype.renderMonthLabel = function (month, year) {
            var _a = this.properties, monthNames = _a.monthNames, renderMonthLabel = _a.renderMonthLabel;
            return renderMonthLabel ? renderMonthLabel(month, year) : monthNames[month].long + " " + year;
        };
        DatePicker.prototype.renderMonthRadios = function () {
            var _this = this;
            var month = this.properties.month;
            return this.properties.monthNames.map(function (monthName, i) { return d_1.v('label', {
                key: _this._idBase + "_month_radios_" + i,
                classes: _this.theme([css.monthRadio, i === month ? css.monthRadioChecked : null])
            }, [
                d_1.v('input', {
                    checked: i === month,
                    classes: _this.theme(css.monthRadioInput),
                    name: _this._idBase + "_month_radios",
                    tabIndex: _this._monthPopupOpen ? 0 : -1,
                    type: 'radio',
                    value: "" + i,
                    onchange: _this._onMonthRadioChange,
                    onmouseup: _this._closeMonthPopup
                }),
                d_1.v('abbr', {
                    classes: _this.theme(css.monthRadioLabel),
                    title: monthName.long
                }, [monthName.short])
            ]); });
        };
        DatePicker.prototype.renderPagingButtonContent = function (type) {
            var labels = this.properties.labels;
            var iconClass = type === "next" /* next */ ? iconCss.rightIcon : iconCss.leftIcon;
            var labelText = type === "next" /* next */ ? labels.nextMonth : labels.previousMonth;
            return [
                d_1.v('i', { classes: this.theme([iconCss.icon, iconClass]),
                    role: 'presentation', 'aria-hidden': 'true'
                }),
                d_1.v('span', { classes: baseCss.visuallyHidden }, [labelText])
            ];
        };
        DatePicker.prototype.renderYearRadios = function () {
            var year = this.properties.year;
            var radios = [];
            var yearLimits = this._getYearRange();
            for (var i = yearLimits.first; i < yearLimits.last; i++) {
                radios.push(d_1.v('label', {
                    key: this._idBase + "_year_radios_" + i,
                    classes: this.theme([css.yearRadio, i === year ? css.yearRadioChecked : null])
                }, [
                    d_1.v('input', {
                        checked: i === year,
                        classes: this.theme(css.yearRadioInput),
                        name: this._idBase + "_year_radios",
                        tabIndex: this._yearPopupOpen ? 0 : -1,
                        type: 'radio',
                        value: "" + i,
                        onchange: this._onYearRadioChange,
                        onmouseup: this._closeYearPopup
                    }),
                    d_1.v('abbr', {
                        classes: this.theme(css.yearRadioLabel)
                    }, ["" + i])
                ]));
            }
            return radios;
        };
        DatePicker.prototype.render = function () {
            var _a = this.properties, _b = _a.labelId, labelId = _b === void 0 ? this._idBase + "_label" : _b, labels = _a.labels, month = _a.month, year = _a.year;
            return d_1.v('div', {
                classes: this.theme(css.datePicker)
            }, [
                d_1.v('div', {
                    classes: this.theme(css.topMatter),
                    role: 'menubar'
                }, [
                    // hidden label
                    d_1.v('label', {
                        id: labelId,
                        classes: [baseCss.visuallyHidden],
                        'aria-live': 'polite',
                        'aria-atomic': 'false'
                    }, [this.renderMonthLabel(month, year)]),
                    // month trigger
                    this.renderControlsTrigger("month" /* month */),
                    // year trigger
                    this.renderControlsTrigger("year" /* year */)
                ]),
                // month grid
                d_1.v('div', {
                    key: 'month-grid',
                    'aria-hidden': "" + !this._monthPopupOpen,
                    'aria-labelledby': this._idBase + "_month_button",
                    classes: [this.theme(css.monthGrid), !this._monthPopupOpen ? baseCss.visuallyHidden : null],
                    id: this._idBase + "_month_dialog",
                    role: 'dialog'
                }, [
                    d_1.v('fieldset', {
                        classes: this.theme(css.monthFields),
                        onkeydown: this._onPopupKeyDown
                    }, tslib_1.__spread([
                        d_1.v('legend', { classes: baseCss.visuallyHidden }, [labels.chooseMonth])
                    ], this.renderMonthRadios()))
                ]),
                // year grid
                d_1.v('div', {
                    key: 'year-grid',
                    'aria-hidden': "" + !this._yearPopupOpen,
                    'aria-labelledby': this._idBase + "_year_button",
                    classes: [this.theme(css.yearGrid), !this._yearPopupOpen ? baseCss.visuallyHidden : null],
                    id: this._idBase + "_year_dialog",
                    role: 'dialog'
                }, [
                    d_1.v('fieldset', {
                        classes: this.theme(css.yearFields),
                        onkeydown: this._onPopupKeyDown
                    }, tslib_1.__spread([
                        d_1.v('legend', { classes: [baseCss.visuallyHidden] }, [labels.chooseYear])
                    ], this.renderYearRadios())),
                    d_1.v('div', {
                        classes: this.theme(css.controls)
                    }, [
                        d_1.v('button', {
                            classes: this.theme(css.previous),
                            tabIndex: this._yearPopupOpen ? 0 : -1,
                            onclick: this._onYearPageDown
                        }, this.renderPagingButtonContent("previous" /* previous */)),
                        d_1.v('button', {
                            classes: this.theme(css.next),
                            tabIndex: this._yearPopupOpen ? 0 : -1,
                            onclick: this._onYearPageUp
                        }, this.renderPagingButtonContent("next" /* next */))
                    ])
                ])
            ]);
        };
        DatePicker = tslib_1.__decorate([
            Themed_1.theme(css),
            Themed_1.theme(iconCss)
        ], DatePicker);
        return DatePicker;
    }(exports.DatePickerBase));
    exports.default = DatePicker;
});
