// @dojo/widgets/calendar/DatePicker.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "@dojo/widget-core/WidgetBase", "@dojo/widget-core/mixins/Themeable", "@dojo/widget-core/d", "@dojo/core/uuid", "./styles/calendar.m.css", "../common/styles/base.m.css", "../common/styles/icons.m.css"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var WidgetBase_1 = require("@dojo/widget-core/WidgetBase");
    var Themeable_1 = require("@dojo/widget-core/mixins/Themeable");
    var d_1 = require("@dojo/widget-core/d");
    var uuid_1 = require("@dojo/core/uuid");
    var css = require("./styles/calendar.m.css");
    var baseCss = require("../common/styles/base.m.css");
    var iconCss = require("../common/styles/icons.m.css");
    ;
    ;
    exports.DatePickerBase = Themeable_1.ThemeableMixin(WidgetBase_1.WidgetBase);
    var BASE_YEAR = 2000;
    var DatePicker = /** @class */ (function (_super) {
        __extends(DatePicker, _super);
        function DatePicker() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this._callMonthTriggerFocus = false;
            _this._callYearTriggerFocus = false;
            _this._callMonthPopupFocus = false;
            _this._callYearPopupFocus = false;
            _this._idBase = uuid_1.default();
            _this._monthPopupOpen = false;
            _this._yearPopupOpen = false;
            _this._yearPage = 0;
            return _this;
        }
        DatePicker.prototype._closeMonthPopup = function () {
            var onPopupChange = this.properties.onPopupChange;
            this._monthPopupOpen = false;
            this._callMonthTriggerFocus = true;
            this.invalidate();
            onPopupChange && onPopupChange(this._getPopupState());
        };
        DatePicker.prototype._closeYearPopup = function () {
            var onPopupChange = this.properties.onPopupChange;
            this._yearPopupOpen = false;
            this._callYearTriggerFocus = true;
            this.invalidate();
            onPopupChange && onPopupChange(this._getPopupState());
        };
        DatePicker.prototype._getPopupState = function () {
            return this._monthPopupOpen || this._yearPopupOpen;
        };
        DatePicker.prototype._getYearRange = function () {
            var _a = this.properties, year = _a.year, _b = _a.yearRange, yearRange = _b === void 0 ? 20 : _b;
            var offset = (year - BASE_YEAR) % yearRange - yearRange * this._yearPage;
            if (year >= BASE_YEAR) {
                return { first: year - offset, last: year + yearRange - offset };
            }
            else {
                return { first: year - (yearRange + offset), last: year - offset };
            }
        };
        // move focus when opening/closing the popup
        DatePicker.prototype.onElementUpdated = function (element, key) {
            if (key === 'month-button') {
                if (!this._monthPopupOpen && this._callMonthTriggerFocus) {
                    element.focus();
                    this._callMonthTriggerFocus = false;
                }
            }
            if (key === 'year-button') {
                if (!this._yearPopupOpen && this._callYearTriggerFocus) {
                    element.focus();
                    this._callYearTriggerFocus = false;
                }
            }
            if (this._callMonthPopupFocus && key.indexOf(this._idBase + "_month_radios") > -1) {
                var month = key.split('_')[3];
                if (this._monthPopupOpen && month === "" + this.properties.month) {
                    element.children[0].focus();
                    this._callMonthPopupFocus = false;
                }
            }
            if (this._callYearPopupFocus && key.indexOf(this._idBase + "_year_radios") > -1) {
                var year = key.split('_')[3];
                if (this._yearPopupOpen && year === "" + this.properties.year) {
                    element.children[0].focus();
                    this._callYearPopupFocus = false;
                }
            }
        };
        DatePicker.prototype._onMonthButtonClick = function () {
            this._monthPopupOpen ? this._closeMonthPopup() : this._openMonthPopup();
        };
        DatePicker.prototype._onMonthRadioChange = function (event) {
            var onRequestMonthChange = this.properties.onRequestMonthChange;
            onRequestMonthChange && onRequestMonthChange(parseInt(event.target.value, 10));
        };
        DatePicker.prototype._onPopupKeyDown = function (event) {
            // close popup on escape, or if a value is selected with enter/space
            if (event.which === 27 /* Escape */ ||
                event.which === 13 /* Enter */ ||
                event.which === 32 /* Space */) {
                this._closeMonthPopup();
                this._closeYearPopup();
            }
        };
        DatePicker.prototype._onYearButtonClick = function () {
            this._yearPopupOpen ? this._closeYearPopup() : this._openYearPopup();
        };
        DatePicker.prototype._onYearPageDown = function () {
            this._yearPage--;
            this._yearPopupOpen && this.invalidate();
        };
        DatePicker.prototype._onYearPageUp = function () {
            this._yearPage++;
            this._yearPopupOpen && this.invalidate();
        };
        DatePicker.prototype._onYearRadioChange = function (event) {
            var onRequestYearChange = this.properties.onRequestYearChange;
            this._yearPage = 0;
            onRequestYearChange && onRequestYearChange(parseInt(event.target.value, 10));
        };
        DatePicker.prototype._openMonthPopup = function () {
            var onPopupChange = this.properties.onPopupChange;
            this._monthPopupOpen = true;
            this._callMonthPopupFocus = true;
            this._yearPopupOpen = false;
            this.invalidate();
            onPopupChange && onPopupChange(this._getPopupState());
        };
        DatePicker.prototype._openYearPopup = function () {
            var onPopupChange = this.properties.onPopupChange;
            this._yearPopupOpen = true;
            this._callYearPopupFocus = true;
            this._monthPopupOpen = false;
            this.invalidate();
            onPopupChange && onPopupChange(this._getPopupState());
        };
        DatePicker.prototype._renderMonthRadios = function () {
            var _this = this;
            var month = this.properties.month;
            return this.properties.monthNames.map(function (monthName, i) { return d_1.v('label', {
                key: _this._idBase + "_month_radios_" + i,
                classes: _this.classes(css.monthRadio, i === month ? css.monthRadioChecked : null)
            }, [
                d_1.v('input', {
                    checked: i === month,
                    classes: _this.classes(css.monthRadioInput),
                    name: _this._idBase + "_month_radios",
                    tabIndex: _this._monthPopupOpen ? 0 : -1,
                    type: 'radio',
                    value: "" + i,
                    onchange: _this._onMonthRadioChange,
                    onmouseup: _this._closeMonthPopup
                }),
                d_1.v('abbr', {
                    classes: _this.classes(css.monthRadioLabel),
                    title: monthName.long
                }, [monthName.short])
            ]); });
        };
        DatePicker.prototype._renderYearRadios = function () {
            var year = this.properties.year;
            var radios = [];
            var yearLimits = this._getYearRange();
            for (var i = yearLimits.first; i < yearLimits.last; i++) {
                radios.push(d_1.v('label', {
                    key: this._idBase + "_year_radios_" + i,
                    classes: this.classes(css.yearRadio, i === year ? css.yearRadioChecked : null)
                }, [
                    d_1.v('input', {
                        checked: i === year,
                        classes: this.classes(css.yearRadioInput),
                        name: this._idBase + "_year_radios",
                        tabIndex: this._yearPopupOpen ? 0 : -1,
                        type: 'radio',
                        value: "" + i,
                        onchange: this._onYearRadioChange,
                        onmouseup: this._closeYearPopup
                    }),
                    d_1.v('abbr', {
                        classes: this.classes(css.yearRadioLabel)
                    }, ["" + i])
                ]));
            }
            return radios;
        };
        DatePicker.prototype._renderMonthLabel = function (month, year) {
            var _a = this.properties, monthNames = _a.monthNames, renderMonthLabel = _a.renderMonthLabel;
            return renderMonthLabel ? renderMonthLabel(month, year) : monthNames[month].long + " " + year;
        };
        DatePicker.prototype.render = function () {
            var _a = this.properties, _b = _a.labelId, labelId = _b === void 0 ? this._idBase + "_label" : _b, labels = _a.labels, month = _a.month, monthNames = _a.monthNames, year = _a.year;
            return d_1.v('div', {
                classes: this.classes(css.datePicker)
            }, [
                d_1.v('div', {
                    classes: this.classes(css.topMatter),
                    role: 'menubar'
                }, [
                    // hidden label
                    d_1.v('label', {
                        id: labelId,
                        classes: this.classes().fixed(baseCss.visuallyHidden),
                        'aria-live': 'polite',
                        'aria-atomic': 'false'
                    }, [this._renderMonthLabel(month, year)]),
                    // month trigger
                    d_1.v('button', {
                        key: 'month-button',
                        'aria-controls': this._idBase + "_month_dialog",
                        'aria-expanded': "" + this._monthPopupOpen,
                        'aria-haspopup': 'true',
                        id: this._idBase + "_month_button",
                        classes: this.classes(css.monthTrigger, this._monthPopupOpen ? css.monthTriggerActive : null),
                        role: 'menuitem',
                        onclick: this._onMonthButtonClick
                    }, [monthNames[month].long]),
                    // year trigger
                    d_1.v('button', {
                        key: 'year-button',
                        'aria-controls': this._idBase + "_year_dialog",
                        'aria-expanded': "" + this._yearPopupOpen,
                        'aria-haspopup': 'true',
                        id: this._idBase + "_year_button",
                        classes: this.classes(css.yearTrigger, this._yearPopupOpen ? css.yearTriggerActive : null),
                        role: 'menuitem',
                        onclick: this._onYearButtonClick
                    }, ["" + year])
                ]),
                // month grid
                d_1.v('div', {
                    key: 'month-grid',
                    'aria-hidden': "" + !this._monthPopupOpen,
                    'aria-labelledby': this._idBase + "_month_button",
                    classes: this.classes(css.monthGrid).fixed(!this._monthPopupOpen ? baseCss.visuallyHidden : null),
                    id: this._idBase + "_month_dialog",
                    role: 'dialog'
                }, [
                    d_1.v('fieldset', {
                        classes: this.classes(css.monthFields),
                        onkeydown: this._onPopupKeyDown
                    }, [
                        d_1.v('legend', { classes: this.classes().fixed(baseCss.visuallyHidden) }, [labels.chooseMonth])
                    ].concat(this._renderMonthRadios()))
                ]),
                // year grid
                d_1.v('div', {
                    key: 'year-grid',
                    'aria-hidden': "" + !this._yearPopupOpen,
                    'aria-labelledby': this._idBase + "_year_button",
                    classes: this.classes(css.yearGrid).fixed(!this._yearPopupOpen ? baseCss.visuallyHidden : null),
                    id: this._idBase + "_year_dialog",
                    role: 'dialog'
                }, [
                    d_1.v('fieldset', {
                        classes: this.classes(css.yearFields),
                        onkeydown: this._onPopupKeyDown
                    }, [
                        d_1.v('legend', { classes: this.classes().fixed(baseCss.visuallyHidden) }, [labels.chooseYear])
                    ].concat(this._renderYearRadios())),
                    d_1.v('div', {
                        classes: this.classes(css.controls)
                    }, [
                        d_1.v('button', {
                            classes: this.classes(css.previous),
                            tabIndex: this._yearPopupOpen ? 0 : -1,
                            onclick: this._onYearPageDown
                        }, [
                            d_1.v('i', { classes: this.classes(iconCss.icon, iconCss.leftIcon),
                                role: 'presentation', 'aria-hidden': 'true'
                            }),
                            d_1.v('span', { classes: this.classes().fixed(baseCss.visuallyHidden) }, [labels.previousMonth])
                        ]),
                        d_1.v('button', {
                            classes: this.classes(css.next),
                            tabIndex: this._yearPopupOpen ? 0 : -1,
                            onclick: this._onYearPageUp
                        }, [
                            d_1.v('i', { classes: this.classes(iconCss.icon, iconCss.rightIcon),
                                role: 'presentation', 'aria-hidden': 'true'
                            }),
                            d_1.v('span', { classes: this.classes().fixed(baseCss.visuallyHidden) }, [labels.nextMonth])
                        ])
                    ])
                ])
            ]);
        };
        DatePicker = __decorate([
            Themeable_1.theme(css),
            Themeable_1.theme(iconCss)
        ], DatePicker);
        return DatePicker;
    }(exports.DatePickerBase));
    exports.default = DatePicker;
});
