// @dojo/widget-core/mixins/I18n.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "@dojo/core/lang", "@dojo/i18n/i18n", "./../WidgetBase", "./../d"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    /* tslint:disable:interface-name */
    var lang_1 = require("@dojo/core/lang");
    var i18n_1 = require("@dojo/i18n/i18n");
    var WidgetBase_1 = require("./../WidgetBase");
    var d_1 = require("./../d");
    function I18nMixin(Base) {
        var I18n = (function (_super) {
            __extends(I18n, _super);
            function I18n() {
                var args = [];
                for (var _i = 0; _i < arguments.length; _i++) {
                    args[_i] = arguments[_i];
                }
                var _this = _super.apply(this, args) || this;
                var subscription = i18n_1.observeLocale({
                    next: function () {
                        if (!_this.properties.locale) {
                            _this.invalidate();
                        }
                    }
                });
                _this.own({
                    destroy: function () {
                        subscription.unsubscribe();
                    }
                });
                return _this;
            }
            I18n.prototype.localizeBundle = function (bundle) {
                var locale = this.properties.locale;
                var messages = this._getLocaleMessages(bundle) || bundle.messages;
                return lang_1.assign(Object.create({
                    format: function (key, options) {
                        return i18n_1.formatMessage(bundle.bundlePath, key, options, locale);
                    }
                }), messages);
            };
            I18n.prototype.renderDecorator = function (result) {
                if (d_1.isHNode(result)) {
                    var _a = this.properties, locale = _a.locale, rtl = _a.rtl;
                    var vNodeProperties = {
                        dir: null,
                        lang: null
                    };
                    if (typeof rtl === 'boolean') {
                        vNodeProperties['dir'] = rtl ? 'rtl' : 'ltr';
                    }
                    if (locale) {
                        vNodeProperties['lang'] = locale;
                    }
                    lang_1.assign(result.properties, vNodeProperties);
                }
                return result;
            };
            /**
             * @private
             * Return the cached dictionary for the specified bundle and locale, if it exists. If the requested dictionary does not
             * exist, then load it and update the instance's state with the appropriate messages.
             *
             * @param bundle
             * The bundle for which to load a locale-specific dictionary.
             *
             * @return
             * The locale-specific dictionary, if it has already been loaded and cached.
             */
            I18n.prototype._getLocaleMessages = function (bundle) {
                var _this = this;
                var properties = this.properties;
                var locale = properties.locale || i18n_1.default.locale;
                var localeMessages = i18n_1.getCachedMessages(bundle, locale);
                if (localeMessages) {
                    return localeMessages;
                }
                i18n_1.default(bundle, locale).then(function () {
                    _this.invalidate();
                });
            };
            return I18n;
        }(Base));
        __decorate([
            WidgetBase_1.afterRender(),
            __metadata("design:type", Function),
            __metadata("design:paramtypes", [Object]),
            __metadata("design:returntype", Object)
        ], I18n.prototype, "renderDecorator", null);
        return I18n;
    }
    exports.I18nMixin = I18nMixin;
    exports.default = I18nMixin;
});
