// @dojo/core/request/Response.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "@dojo/shim/Promise", "../QueuingEvented"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var Promise_1 = require("@dojo/shim/Promise");
    var QueuingEvented_1 = require("../QueuingEvented");
    var Response = (function (_super) {
        __extends(Response, _super);
        function Response() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        Response.prototype.json = function () {
            return this.text().then(JSON.parse);
        };
        return Response;
    }(QueuingEvented_1.default));
    exports.default = Response;
    function getFileReaderPromise(reader) {
        return new Promise_1.default(function (resolve, reject) {
            reader.onload = function () {
                resolve(reader.result);
            };
            reader.onerror = function () {
                reject(reader.error);
            };
        });
    }
    exports.getFileReaderPromise = getFileReaderPromise;
    function getTextFromBlob(blob) {
        var reader = new FileReader();
        var promise = getFileReaderPromise(reader);
        reader.readAsText(blob);
        return promise;
    }
    exports.getTextFromBlob = getTextFromBlob;
    function getArrayBufferFromBlob(blob) {
        var reader = new FileReader();
        var promise = getFileReaderPromise(reader);
        reader.readAsArrayBuffer(blob);
        return promise;
    }
    exports.getArrayBufferFromBlob = getArrayBufferFromBlob;
    function getTextFromArrayBuffer(buffer) {
        var view = new Uint8Array(buffer);
        var chars = [];
        view.forEach(function (charCode, index) {
            chars[index] = String.fromCharCode(charCode);
        });
        return chars.join('');
    }
    exports.getTextFromArrayBuffer = getTextFromArrayBuffer;
});
