// @dojo/core/List.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "@dojo/shim/iterator", "@dojo/shim/WeakMap"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var iterator_1 = require("@dojo/shim/iterator");
    var WeakMap_1 = require("@dojo/shim/WeakMap");
    var listItems = new WeakMap_1.default();
    function getListItems(list) {
        return (listItems.get(list) || []);
    }
    var List = (function () {
        function List(source) {
            var _this = this;
            listItems.set(this, []);
            if (source) {
                iterator_1.forOf(source, function (item) {
                    _this.add(item);
                });
            }
        }
        List.prototype[Symbol.iterator] = function () {
            return this.values();
        };
        Object.defineProperty(List.prototype, "size", {
            get: function () {
                return getListItems(this).length;
            },
            enumerable: true,
            configurable: true
        });
        List.prototype.add = function (value) {
            getListItems(this).push(value);
            return this;
        };
        List.prototype.clear = function () {
            listItems.set(this, []);
        };
        List.prototype.delete = function (idx) {
            if (idx < this.size) {
                getListItems(this).splice(idx, 1);
                return true;
            }
            return false;
        };
        List.prototype.entries = function () {
            return new iterator_1.ShimIterator(getListItems(this).map(function (value, index) { return [index, value]; }));
        };
        List.prototype.forEach = function (fn, thisArg) {
            getListItems(this).forEach(fn.bind(thisArg ? thisArg : this));
        };
        List.prototype.has = function (idx) {
            return this.size > idx;
        };
        List.prototype.includes = function (value) {
            return getListItems(this).indexOf(value) >= 0;
        };
        List.prototype.indexOf = function (value) {
            return getListItems(this).indexOf(value);
        };
        List.prototype.join = function (separator) {
            if (separator === void 0) { separator = ','; }
            return getListItems(this).join(separator);
        };
        List.prototype.keys = function () {
            return new iterator_1.ShimIterator(getListItems(this).map(function (_, index) { return index; }));
        };
        List.prototype.lastIndexOf = function (value) {
            return getListItems(this).lastIndexOf(value);
        };
        List.prototype.push = function (value) {
            this.add(value);
        };
        List.prototype.pop = function () {
            return getListItems(this).pop();
        };
        List.prototype.splice = function (start, deleteCount) {
            var newItems = [];
            for (var _i = 2; _i < arguments.length; _i++) {
                newItems[_i - 2] = arguments[_i];
            }
            return (_a = getListItems(this)).splice.apply(_a, [start,
                deleteCount === undefined ? (this.size - start) : deleteCount].concat(newItems));
            var _a;
        };
        List.prototype.values = function () {
            return new iterator_1.ShimIterator(getListItems(this).map(function (value) { return value; }));
        };
        return List;
    }());
    exports.default = List;
});
