// @dojo/widgets/tabpane/createTabElement.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "./Tab"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var Tab_1 = require("./Tab");
    /**
     * Configures a Tab web component
     */
    function createTabElement() {
        return {
            tagName: 'dojo-tab',
            widgetConstructor: Tab_1.default,
            attributes: [
                {
                    attributeName: 'closeable',
                    value: function (value) { return value === 'false' || value === '0' ? false : true; }
                },
                {
                    attributeName: 'disabled',
                    value: function (value) { return value === 'false' || value === '0' ? false : true; }
                },
                {
                    attributeName: 'key'
                }
            ],
            properties: [
                {
                    propertyName: 'label'
                }
            ]
        };
    }
    exports.default = createTabElement;
    ;
});
