// @dojo/widgets/tabpane/createTabPaneElement.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "./TabPane"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var TabPane_1 = require("./TabPane");
    /**
     * Configures a TabPane web component
     */
    function createTabPaneElement() {
        return {
            tagName: 'dojo-tab-pane',
            widgetConstructor: TabPane_1.default,
            attributes: [
                {
                    attributeName: 'activeindex',
                    propertyName: 'activeIndex',
                    value: function (value) { return Number(value); }
                },
                {
                    attributeName: 'alignbuttons',
                    propertyName: 'alignButtons',
                    value: function (value) { return Number(value); }
                }
            ],
            events: [
                {
                    propertyName: 'onRequestTabChange',
                    eventName: 'tabchange'
                },
                {
                    propertyName: 'onRequestTabClose',
                    eventName: 'tabclose'
                }
            ]
        };
    }
    exports.default = createTabPaneElement;
    ;
});
