// @dojo/routing/main.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "./Route", "./Router", "./history/HashHistory", "./history/MemoryHistory", "./history/StateHistory"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var Route_1 = require("./Route");
    exports.Route = Route_1.default;
    var Router_1 = require("./Router");
    exports.Router = Router_1.default;
    var HashHistory_1 = require("./history/HashHistory");
    var MemoryHistory_1 = require("./history/MemoryHistory");
    var StateHistory_1 = require("./history/StateHistory");
    exports.history = {
        HashHistory: HashHistory_1.default,
        MemoryHistory: MemoryHistory_1.default,
        StateHistory: StateHistory_1.default
    };
});
